### PR TITLE
upload blobs in attach()

### LINKED ISSUE #6523: Implement redirection for detached blob URLs
This is a sub task of **_Offline blob uploads Phase 2 - Support "offline" blob uploads with detached container_** #6128

**Approach**
- Implement hook for uploading blobs on Container.attach()
- As detached blobs upload, add their before/after URLs to a table and reclaim detached storage.
- Use the table to resolve requests for "before" URLs.
- Add snapshot/load APIs to IDetachedBlobStorage
- Include blob redirection table in summary.

### PR DESCRIPTION
**Pull request by wes-carlson**
#6128
close #6523
Upload attachment blobs in `attach()` and set the redirect table on `BlobManager`.

### CODE REVIEW

File: api-report/container-loader.api.md
Comment (vladsud) [inline api-report/container-loader.api.md:286]: The shape does not look right, assuming 400Mb of payload I've heard being quoted for these kinds of workloads.
I think it has to be enumeration and it has to be async. That might create other problems (async part of it), but I'm not sure duplicating that much memory is scalable solution.

Comment (vladsud) [reply api-report/container-loader.api.md:286]: Oh, these are IDs? Sorry, misunderstood, I'd change the name and add comment.

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:850]: If this fails, do we delete file? I think host would not have any file identification, so it can't delete this file. Something should have ability to clean it, we should not leave garbage.

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:839]: It would be great to add some comments describing why this flow is needed. I.e. what kind of problem it solves.

Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:839]: I believe you can start with empty redirectTable - loop below will fill it in.

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:842]: Can we do all() call less frequently? Like at the end of the loop and only if there are no more blobs to work with?

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:858]: I do not have good suggestion, but we should brainstorm if there is a way to remove the need to have this API
Maybe by moving some of the logic here to container runtime (unlikely as it would add dependency on something else)? Or maybe this mapping should stay at loader level and loader writes out mapping blob on each summary?
It just looks weird from contract perspective, especially if scenarios does not create attachment blobs in detached container.

Comment (vladsud) [reply packages/loader/container-loader/src/container.ts:858]: Alternatively it should be a payload on createSummary() function. That seems a bit more logical IMHO (though not sure if createSummary() is called outside of attachment flow)

File: packages/loader/container-loader/src/containerContext.ts
Comment (vladsud) [inline packages/loader/container-loader/src/containerContext.ts:281]: This casting should be fixed by changing definitions. Also it should be optional (and container fail if it's not there with some more reasonable error message than just "can't access setBlobRedirectTable on undefined").
We could make it non-optional, but it seems a bit against our principles on how we keep changes between loader & container runtime (even though it is for scenario that does not yet exists)

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:88]: I do not think there is big difference in order, so I'd move usage of this.redirectTable up and remove recursion in this function. I.e.

```typescript
let storageId = this.redirectTable?.get(blobId);
if (storageId  === undefined) {
   storageId = blobId;
}

assert(this.blobIds.has(blobId) || this.pendingBlobIds.has(blobId) || this.detachedBlobIds.has(blobId), 0x11f /* "requesting unknown blobs" */);
```

Also is there any way to reduce number of concepts? Is there a way to merge pendingBlobIds & detachedBlobIds (as they can't be used at the same time, and they have roughly same meaning - blobs that are in progress).

**Comment by vladsud** [comment]
@DLehenbauer - please glance as well, maybe you have good ideas on the loader - container runtime API and reduction of concepts - telling about redirection table is not great, though I do not see better option at the moment.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/blobManager.ts]: I don't think it would make sense to merge these. detachedBlobIds could be removed since it's only used to validate the redirect table, which is probably not really necessary.

File: packages/loader/container-loader/src/containerContext.ts
Comment (vladsud) [inline packages/loader/container-loader/src/containerContext.ts:206]: It would be great to avoid casting here. I assume container definition needs to be changed first and versions bumped to make it happen, right?

File: packages/loader/container-loader/src/loader.ts
Comment (vladsud) [inline packages/loader/container-loader/src/loader.ts:268]: Maybe getBlobIds()? Interface may include more things in the future, so it's useful for method names to be more descriptive

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:98]: Consider adding telemetry event here such that we know how often it happens and either have a chance to revisit this area or advice customer not to change their design

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:61]: I do not see redirectTable being summarized. I believe this makes it impossible for client loading from snapshot to resolve blob handles. Is this covered in UTs?

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:80]: I do not see detached blobs being added to this.blobIds(), and thus making logic inconsistent. For example, blobCount() would not properly count them. I also think you should not check this.pendingBlobIds, as there should be no way for callers to have such blobId until it lands into this.blobIds. I think same is true for this.redirectTable (for somewhat similar reasons)

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:101]: Any reason not to use redirection table all the time? I.e. even in attached state, add a mapping (identity mapping is fine) such that all the code is consistent in how it resolves IDs? this.blobIds would likely not be needed, as this.redirectTable would be source of truth.
For detached -> attaching transition, we can simply patch this.redirectTable, but leave otherwise most of the assumptions exactly the same.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/blobManager.ts:80]: They are added in setRedirectTable(), but you are right about this.pendingBlobIds/this.redirectTable

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/blobManager.ts:61]: This is not implemented in this PR. Loading from snapshot (after creating empty file) is not working on ODSP and other drivers don't support empty file creation.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/blobManager.ts:101]: Since the redirect table needs to be stored in snapshot, I think we want it kept separate from this.blobIds.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (DLehenbauer) [reply packages/runtime/container-runtime/src/blobManager.ts:61]: Still true, or did avoiding the file rename unblock this work?

File: packages/loader/container-loader/src/container.ts
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:895]: Consider wrapping this whole process with try/catch and adding resolved URI on error object in some way, such that host has an ability to find this document and delete it.

File: packages/loader/container-loader/src/containerContext.ts
Comment (vladsud) [inline packages/loader/container-loader/src/containerContext.ts:214]: Can you please add comment (ideally here and in definitions) on the expectations and meaning of the argument?

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobManager.ts:61]: I meant in the first real summary that we produce as part of this flow. redirectTable has to be serialized in summary in some form or shape to be able to map from internal IDs (coming from detached storage) to real storage IDs.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:159]: I'd appreciate adding this property to existing event above and not create new events.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobManager.ts:61]: And I do see that you serialize it below, I think it was not the case when I left this comment

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:171]: Do we have UTs for these flows? If not, it should be a requirement for finishing this work.

File: packages/runtime/container-runtime/src/blobManager.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/blobManager.ts:176]: I believe this API should be allowed to be called only once and only in specific state of container. Can we assert all these pre-conditions?

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1218]: Can all of this logic (RE what is in the blob manager summary) be in blob manager layer, and this layer not caring about details?

**Comment by vladsud** [comment]
Looks good overall, please take a look at comments. Thanks!

File: packages/runtime/container-runtime/src/containerRuntime.ts
Comment (wes-carlson) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1218]: I think the only way to do this would be to move BlobManager loading out of the ContainerRuntime constructor into ContainerRuntime.load(), since the attachment types don't need to be read from storage but other blobs do.

Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1218]: I think we should definitely consider that in general. I.e. move some of the logic in ContainerRuntime.load() and ContainerRuntime.constructor into async ContainerRuntime.initialize() and call it from load() as part of the loading flow.

Comment (wes-carlson) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: I've done this, please take a look

File: api-report/container-runtime-definitions.api.md
Comment (tylerbutler) [inline api-report/container-runtime-definitions.api.md:73]: Is this a safe change with respect to back-compat? Are there any scenarios where developers _must_ handle the "attached" event?

Comment (wes-carlson) [reply api-report/container-runtime-definitions.api.md:73]: No. This is only used within ContainerRuntime, so back compat is not an issue.

**Review by vladsud — APPROVED** [review]

**Pull request merged**
